[WIP] Add credential plugin allowlist to kuberc

pmengelbert opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Per [a proposed update to KEP 3104](https://github.com/kubernetes/enhancements/pull/5479), this PR implements an allowlist for client-go credential plugins. The allowlist is configured in the kuberc configuration file. This feature improves security by giving the user control over which binaries are allowed to execute in response to the configured credential plugins in the kubeconfig.

An unspecified/`nil` allowlist will permit the execution of any binary as a credential plugin (thus maintaining backward compatibility). An explicitly empty allowlist will not permit any binaries as credential plugins. Please see the KEP update PR, linked above, for details on exactly how the allowlist arrives at a positive decision to run a binary.

#### Which issue(s) this PR is related to:
- KEP issue: https://github.com/kubernetes/enhancements/issues/3104
- KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/3104-introduce-kuberc/README.md
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

pmengelbert committed:
Add validation for credential plugin allowlist yaml

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Implement credential plugin allowlist

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Remove dead code

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Remove redundant error wrapping

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Align empty string check with kube conventions

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Add error log

When `exec.LookPath` fails on one of the binary names in the allowlist,
log the error but do not fail the operation.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Add emptiness check for allowlist entries

If an allowlist entry is entirely empty, it should be considered a noop,
and should not allow any binaries to run.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Add doc comments for new fields/types in kuberc

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Move allowlist validation after other checks

I made a logical error here. The way the kuberc decoder is designed
permits multiple versioned instances of a full kuberc yaml document in
the same file. If lenient decoding fails, the next document will be
tried.

The way my code was structured prevented subsequent yaml documents from
being tried. If the first kuberc document in the file failed strict
decoding, but another kuberc document in the file would have passed both
strict and lenient decoding, my code would prevent a valid kuberc from
being used.

The allowlist validation check should also take place after the checks
on the yaml document's GroupKind.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Update docs for relevant fields and structs

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Be more specific in doc comment

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Remove manual emptiness check

Instead, compare against the zero value for `AllowlistItem`.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Update docs comments

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Decouple kubeconfig and kuberc

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

benjaminapetersen commented on staging/src/k8s.io/kubectl/pkg/config/v1alpha1/types.go:
```diff
@@ -62,6 +62,34 @@ type Preference struct {
 	// +listType=atomic
 	Aliases []AliasOverride `json:"aliases"`
+
+	// credPluginAllowlist (the credential plugin allowlist) specifies the
```
Nit: caps the `c` in `credPluginAllowList`.

pmengelbert commented:
The kuberc is currently being read in the code for the `kubectl` root-level command itself, but as far as I can tell the `Preference`s are discarded after `.Apply(...)` is called. Is there a way to access the preferences from the subcommands without going through the unmarshaling process again? And if we have to push the allowlist check down into the subcommands, is there a way to avoid having to add the check to every subcommand that needs it?

pmengelbert committed:
Improve error messages and comments

- The plugin allowlist error was not specific enough.
- The reasoning for causing failure needed to be explained more clearly,
  because it deviates from the prevailing pattern in the existing
  implementation.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Add special error message when allowlist is empty

When the allowlist is explicitly the empty list, make sure that the user
understands the behavior (all plugins are denied).

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>
